tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,1173 @@
+    assert_deps(deps, {"item": Product})
+
+    # calling the actual page object should also work
+    item, deps = yield crawl_item_and_deps(ProductPage)
+    assert item == Product(name="product name")
+    assert_deps(deps, {"page": ProductPage})
+
+
+@attrs.define
+class ProductFromParent:
+    name: str
+
+
+@handle_urls(URL)
+class ParentProductPage(ItemPage[ProductFromParent]):
+    @field
+    def name(self) -> str:
+        return "parent product name"
+
+
+@handle_urls(URL)
+class SubclassProductPage(ParentProductPage):
+    @field
+    def name(self) -> str:
+        return "subclass product name"
+
+
+@inlineCallbacks
+def test_item_return_subclass() -> None:
+    """A page object should properly derive the ``Return[ItemType]`` that it
+    inherited from its parent.
+
+    In this test case, there's a clash for the ``url_matcher.Patterns`` since
+    they're exactly the same. This produces a warning message. For conflicts
+    like this, scrapy-poet follows the first ``ApplyRule`` it finds inside the
+    ``SCRAPY_POET_OVERRIDES`` setting.
+
+    To remove this warning, the user should update the priority in
+    ``url_matcher.Patterns`` which is set in ``ApplyRule.for_patterns``.
+    """
+
+    # There should be a warning to the user about clashing rules.
+    rules = [
+        ApplyRule(URL, use=ParentProductPage, to_return=ProductFromParent),
+        ApplyRule(URL, use=SubclassProductPage, to_return=ProductFromParent),
+    ]
+    msg = f"Consider updating the priority of these rules: {rules}"
+
+    with warnings.catch_warnings(record=True) as caught_warnings:
+        item, deps = yield crawl_item_and_deps(ProductFromParent)
+        assert any([True for w in caught_warnings if msg in str(w.message)])
+
+    assert item == ProductFromParent(name="subclass product name")
+    assert_deps(deps, {"item": ProductFromParent})
+
+    # calling the actual page objects should still work
+
+    item, deps = yield crawl_item_and_deps(ParentProductPage)
+    assert item == ProductFromParent(name="parent product name")
+    assert_deps(deps, {"page": ParentProductPage})
+
+    item, deps = yield crawl_item_and_deps(SubclassProductPage)
+    assert item == ProductFromParent(name="subclass product name")
+    assert_deps(deps, {"page": SubclassProductPage})
+
+
+@attrs.define
+class PriorityProductFromParent:
+    name: str
+
+
+@handle_urls(URL, priority=600)
+class PriorityParentProductPage(ItemPage[PriorityProductFromParent]):
+    @field
+    def name(self) -> str:
+        return "priority parent product name"
+
+
+@handle_urls(URL)
+class PrioritySubclassProductPage(PriorityParentProductPage):
+    @field
+    def name(self) -> str:
+        return "priority subclass product name"
+
+
+@inlineCallbacks
+def test_item_return_subclass_priority() -> None:
+    """Same case as with ``test_item_return_subclass()`` but now the parent PO
+    uses a higher priority of 600 than the default 500.
```

Comment:
Should we check the case where the subclass priority is set higher? It seems that's the "happy path", while using a high priority in the base class is an edge case, with little practical application.

Reply:
https://github.com/scrapinghub/scrapy-poet/pull/88/commits/ef01f11591242309802136185ee0fd5bbe15be4f